### ISSUE TITLE
 Questions about the inference process

### ISSUE THREAD

**Issue by JejuWayfarer** [UKPLab/sentence-transformers#354]
Hi, I have several questions.
1. When comparing the similarity of 10000 sentences in the paper, BERT takes ~65h and SBERT takes ~5s.
As I understand it, in the case of BERT, Cosine Similarity is obtained by putting two sentences in the BERT layer during the inference process. In the case of SBERT, I also understood that the Cosine Similarity is obtained by putting two sentences in the SBERT layer(In the picture below).
![image](https://user-images.githubusercontent.com/49282663/90094380-c0d8ea80-dd68-11ea-962a-a5574ab7fe81.png)

2.In such a process, I don't understand that SBERT is much faster.
In another issue article, SBERT does not use Siamese layers in the inference process, but I don't understand this.
Also, In the case of SBERT, I am curious about how to create the final embedding.

3. When fine-learning STSb data, is it correct to learn the weight of SBERT with a regression objective function?

Thanks for reading:)

**Comment by nreimers** [comment]
Hi @JejuWayfarer 
When you cluster 10k examples with BERT, you need to compute the score for 10k*(10k-1)/2 = 50 Million sentence combinations.

Not that in the picture the two BERT+pooling layers are identical. So if you want to cluster 10k examples, you compute the embedding for the 10k sentences and you get an embedding for each. You don't have to use that siamese structure, you only use BERT+pooling for each individual sentence. The siamese structure is just there to depict how the model is fine-tuned.

Then you only need to compute the cosine similarity between all possible 10k sentences. However, computing cosine similarity is quite efficient and you only need a few Milliseconds to get the cosine-similarity between all combinations for your 10k sentences.

In conclusion: With a BERT cross-encoder, you need to pass 50 Million sentence pairs to BERT
With SBERT, you pass 10k sentences to BERT and then use cosine-similarity to compute all pairwise distances (takes few milli seconds.


The final embedding of SBERT is generated by mean pooling the last layer of BERT, i.e., the output embeddings for each token.

Yes, on STSb, a regression objective is used. The following loss is minimized: ||gold_label - cosine_sim(u,v)||_2

Best
Nils Reimers

**Comment by JejuWayfarer** [comment]
Thank you so much for your kind answer.
I have an understanding of how BERT and SBERT work.
I have additional questions

1. If I use mean pooling of the output layer of BERT to find the similarity, is the speed the same as SBERT?

2. In the case of No. 1, I confirmed through the results of the paper that SBERT has superior performance compared to BERT. Is it correct?

3. On the other hand, if50 million pairs passed through bert, it takes very long (~65h), but is it correct that the performance is better than the STS-b result of SBERT?

**Comment by nreimers** [comment]
Hi @JejuWayfarer 
1) Yes, the speed is the same. But as shown in the paper, BERT out-of-the-box produces rather bad sentence embeddings. Fine-tuning is required to get good (meaningful) representations.

2) Correct, SBERT outperforms out-of-the-box BERT with mean pooling by a large margin. BERT was not designed to produce good embeddings.

3) Yes, it is called cross-encoder if you pass both sentences to BERT. A bi-encoder is when you pass sentence individually to the network to get independent embeddings, then are then compared with cosine similarity. Cross-encoder produce often better scores than bi-encoders, as they can look at both inputs simultanously. So if you spend the 65h, your results will be better. In the paper you can find how much better BERT is vs. SBERT.

4) Sadly I don't understand your question. SBERT can be fine-tuned on STSb, by passing the sentences through BERT+pooling, comparing the difference between cosine-sim and the gold label. The error is than back-propagated. This fine-tuning leads to better embeddings for the STSb task.

**Comment by JejuWayfarer** [comment]
Thanks to you, I learned a lot.
1~3 was resolved.
Finetuning SBERT for STSb data means that only the BERT weight at the bottom of the inference picture is upgraded, right?

I understood the advantage of SBERT as being able to compare many sentences in real time.
It is assumed that the embedding of large data that passed fine-tuned BERT (on stsb, output layer, mean pooling) is stored. If I can compare the cosine similarity between stored embeddings when new data comes in, is this a recommended method compared to SBERT?

**Comment by nreimers** [comment]
Hi @JejuWayfarer 
Right, only the BERT weights will be updated. The pooling has no weights.

You can pre-compute embeddings for a large set. If new data comes in, you can compute the embedding for that new text snippet and compare the embedding with your existent embeddings.

One such use case is semantic search, where you have a large corpus of maybe Millions of entries. You have computed the embeddings for all of them. For the new input data, you compute the embedding and compare it with the stored embeddings.

For examples on semantic search, see the examples here:
https://github.com/UKPLab/sentence-transformers/tree/master/examples/applications

**Comment by JejuWayfarer** [comment]
I finally understood. Thank you sincerely:)

**Comment by JejuWayfarer** [comment]
When learning about NLI data in the 'bert-base-nli-stsb-mean-tokens' model, 
not only the BERT weight but also the weight in the picture is learned,
and when learning about stsb is in progress, only the BERT weight is learned, right?

![https://s3-us-west-2.amazonaws.com/secure.notion-static.com/8817a53f-bdbf-48dc-9ff2-1a16f46ed343/Untitled.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/8817a53f-bdbf-48dc-9ff2-1a16f46ed343/Untitled.png)

**Comment by nreimers** [comment]
Right

**Comment by JejuWayfarer** [comment]
It was very helpful. I will close this issue.
